go/vt/servenv/grpc_server_auth_spiffe.go

```diff
@@ -0,0 +1,162 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package servenv
+
+import (
+	"context"
+	"crypto/x509"
+	"net/url"
+	"strings"
+
+	"github.com/spf13/pflag"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+
+	"vitess.io/vitess/go/vt/log"
+)
+
+var (
+	// spiffeTrustDomains list of allowed SPIFFE Trust Domains for client SVIDs during authorization
+	spiffeTrustDomains string
+	// SPIFFEAuthPlugin implements AuthPlugin interface
+	_ Authenticator = (*SPIFFEAuthPlugin)(nil)
+)
+
+// The datatype for spiffe auth Context keys
+type spiffeIdKey int
+
+const (
+	// Internal Context key for the authenticated SPIFFE ID
+	spiffeId spiffeIdKey = 0
+)
+
+func registerGRPCServerAuthSPIFFEFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&spiffeTrustDomains, "grpc_auth_spiffe_allowed_trust_domains", spiffeTrustDomains, "List of allowed SPIFFE Trust Domains for client SVIDs (separated by comma).")
+}
+
+// SPIFFEAuthPlugin implements X.509-based SVID for SPIFFE authentication for grpc. It contains an array of trust domains
+// that will be authorized to connect to the grpc server.
+type SPIFFEAuthPlugin struct {
+	spiffeTrustDomains []string
+}
+
+// Authenticate implements Authenticator interface. This method will be used inside a middleware in grpc_server to authenticate
+// incoming requests.
+func (spa *SPIFFEAuthPlugin) Authenticate(ctx context.Context, fullMethod string) (context.Context, error) {
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		return nil, status.Errorf(codes.Unauthenticated, "no peer connection info")
+	}
+	tlsInfo, ok := p.AuthInfo.(credentials.TLSInfo)
+	if !ok {
+		return nil, status.Errorf(codes.Unauthenticated, "not connected via TLS")
+	}
+
+	cert := tlsInfo.State.PeerCertificates[0] // Only check the leaf certificate
+	spiffeIdUrl, ok := validateSVIDCert(cert, spa.spiffeTrustDomains)
+	if !ok {
+		return nil, status.Errorf(codes.Unauthenticated, "client certificate not authorized")
+	}
+
+	log.Infof("SPIFFE auth plugin has authenticated client with SPIFFE ID %v", spiffeId)
+	return newSPIFFEAuthContext(ctx, spiffeIdUrl), nil
+}
+
+// Validates the given certificate as a valid SVID leaf certificate based on
+// https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md#4-constraints-and-usage
+// and https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md#5-validation
+func validateSVIDCert(cert *x509.Certificate, trustedDomains []string) (*url.URL, bool) {
+	issuedForTrustedDomain := false
+	spiffeId := &url.URL{}
+
+	// Leaf SVIDs should have exactly one URI SAN
+	if len(cert.URIs) != 1 {
+		return spiffeId, false
+	}
+
+	if cert.URIs[0].Scheme != "spiffe" {
+		return spiffeId, false
+	}
+
+	for _, trustDomain := range trustedDomains {
+		if cert.URIs[0].Hostname() == trustDomain {
```

Comment:
Just to clarify, this is explicitly only validating the hostnames obviously, but realistically, would you not want to also validate against the path as well?

Typically, this also can be paired with a (excuse me since I don't fully know the terms here), some server to phone home to to ask "is this URI allowed to do this", similar to OAuth.

Without that full implementation, is there any merit into _only_ validating the hostname rather than the full path?

I can see a case of:

`spiffe://example.com/client1`, and `spiffe://example.com/client2` and wanting to allow client1, but not allow client2.

I'm not _entirely_ sure of all the use cases, but I believe in a lot of contexts, this translates to a per-user issuing a certificate for auth, or per-app, and you'd want to restrict to some apps/users, and not the others.

I think explicitly declaring we only validate the hostname, but I would assume there'd be a desire to do more and the implementation here would make that a bit harder to shoehorn in.

Reply:
Also, I'm not an expert on this subject at all, but I tend to refer to Envoy for a golden standard with this configuration, and they seem to cover a lot more options as well as a SPIRE agent. https://spiffe.io/docs/latest/microservices/envoy-x509/readme/

If we're choosing to implement a subset of the feature set, I think that's personally fine, I just worry if we implement a subset, now others might want more.

Reply:
@mattrobenolt these are great questions/points. For SPIFFE, the entire URI forms the SPIFFE ID, which uniquely identifies a particular user/workload, so things like ACLs could be based on that ID. However, a collection of SPIFFE IDs can belong to a single _trust domain_ (which usually corresponds to a CA, but doesn't necessarily have to). Given this, I think, for accepting connections at all, the thing we should validate is the trust domain, much like we're validating the CA that signed a client cert. This is mostly because we'd accept connections from many individual workloads and users all issued for the same trust domain.

The SPIFFE ID makes more sense to use in ACLs, where we might decide that certain SPIFFE IDs (perhaps some applications or users) should be considered writers, while other IDs are readers.

SPIRE is one implementation of the [SPIFFE Workload API](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Workload_API.md) which makes it an _issuer_ of SVIDs. It can attest to the identity of workloads and then provide them the tools to access protected resources and validate the IDs of other workloads. You could think of it as something like cert-manager + let's encrypt. SPIRE agents are used to create and automatically rotate workload certs and provide the CA bundle for verifying both sides of an mTLS connection. SPIRE is _not_ used as an API to query information about other workloads. Further, SPIRE is **not required** to use SPIFFE as long as you have _some_ issuer for SVIDs.